In [1]:
import pandas as pd
import numpy as np

url = "https://raw.githubusercontent.com/sparks5969/advanced_data_analysis/refs/heads/main/raw_train.csv"
raw_data = pd.read_csv(url)
display(raw_data.describe())

,Id,MSSubClass,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,BsmtFinSF1,...,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,MiscVal,MoSold,YrSold,SalePrice
count,1460.000000,1460.000000,1201.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1452.000000,1460.000000,...,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000,1460.000000
mean,730.500000,56.897260,70.049958,10516.828082,6.099315,5.575342,1971.267808,1984.865753,103.685262,443.639726,...,94.244521,46.660274,21.954110,3.409589,15.060959,2.758904,43.489041,6.321918,2007.815753,180921.195890
std,421.610009,42.300571,24.284752,9981.264932,1.382997,1.112799,30.202904,20.645407,181.066207,456.098091,...,125.338794,66.256028,61.119149,29.317331,55.757415,40.177307,496.123024,2.703626,1.328095,79442.502883
min,1.000000,20.000000,21.000000,1300.000000,1.000000,1.000000,1872.000000,1950.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,2006.000000,34900.000000
25%,365.750000,20.000000,59.000000,7553.500000,5.000000,5.000000,1954.000000,1967.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,5.000000,2007.000000,129975.000000
50%,730.500000,50.000000,69.000000,9478.500000,6.000000,5.000000,1973.000000,1994.000000,0.000000,383.500000,...,0.000000,25.000000,0.000000,0.000000,0.000000,0.000000,0.000000,6.000000,2008.000000,163000.000000
75%,1095.250000,70.000000,80.000000,11601.500000,7.000000,6.000000,2000.000000,2004.000000,166.000000,712.250000,...,168.000000,68.000000,0.000000,0.000000,0.000000,0.000000,0.000000,8.000000,2009.000000,214000.000000
max,1460.000000,190.000000,313.000000,215245.000000,10.000000,9.000000,2010.000000,2010.000000,1600.000000,5644.000000,...,857.000000,547.000000,552.000000,508.000000,480.000000,738.000000,15500.000000,12.000000,2010.000000,755000.000000


# Section 2. Data Preprocessing

In this section, we prepare our raw dataset for machine learning. Real-world data is rarely clean or directly usable, so the preprocessing workflow you learn here is essential for building reliable models.

We will walk through each step of transforming the raw Ames Housing dataset into a fully usable modeling dataset. By the end of this section, you will understand not only *what* to do, but *why* each operation matters.

In this section, you will learn how to:

1. **Split the dataset into training and testing sets**  
   Ensuring that all preprocessing decisions are made using the training data, preventing data leakage and enabling reliable model evaluation.

2. **Identify missing values**  
   Quantifying where missingness occurs and determining which variables require special attention.

3. **Drop variables with excessive missingness**  
   Removing features that are too incomplete to be useful (e.g., more than 40% missing).

4. **Separate numerical and categorical variables**  
   Recognizing that different data types require different preprocessing strategies.

5. **Impute missing values**  
   Using appropriate techniques such as median imputation, KNN imputation, or filling categorical missing values with `"None"`.

6. **Identify ordinal vs. nominal categorical variables**  
   Distinguishing between ordered category scales (e.g., quality ratings) and unordered labels (e.g., neighborhood names).

7. **Encode ordinal variables manually**  
   Mapping ordered categories (e.g., `"Po" < "Fa" < "TA" < "Gd" < "Ex"`) to meaningful integers.

8. **Apply one-hot encoding to nominal variables**  
   Converting unordered categorical labels into binary indicator variables suitable for machine learning models.

Together, these steps form a complete and replicable preprocessing pipeline that you can apply to any structured dataset in future projects.


## Task 4. Split the data into train and test
- Intention: Before performing any preprocessing or model training, we must first split the data into a training set (used to fit preprocessing steps and models) and a test set (used only for evaluation).
Splitting early prevents data leakage and ensures that all learned preprocessing steps (imputers, encoders, scalers) will be fit using only the training data.
- Prompt: Working with the DataFrame called `raw_data` which contains all features plus the target variable `SalePrice`.

1. Separate the dataset into:
   - **X**: all feature columns (raw_data without the `SalePrice` column)
   - **y**: the target column (`SalePrice`)
2. Use `sklearn.model_selection.train_test_split` to split `X` and `y` into:
   - `X_train`, `X_test`
   - `y_train`, `y_test`
3. Use a test size of **20%** and a fixed `random_state=42` for reproducibility.
4. Print:
   - Confirmation that the split was successful
   - The shapes of `X_train`, `X_test`, `y_train`, and `y_test`


In [2]:
from sklearn.model_selection import train_test_split

# 1. Separate the dataset into X and y
X = raw_data.drop(columns=['SalePrice'])
y = raw_data['SalePrice']

# 2. & 3. Split into train and test sets with 20% test size and random_state=42
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 4. Print confirmation and shapes
print("Data split was successful!")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Data split was successful!
X_train shape: (1168, 80)
X_test shape: (292, 80)
y_train shape: (1168,)
y_test shape: (292,)


## In Task 4.2 to 4.10, we will handling missing values

## Task 4.2. check and display the percentage of missing values in each variable
- Intention: we need to have a sense of what variables have lots of missing, and what variable are ready to use
- Prompt: Work with the DataFrame `X_train` and perform the following steps:

1. Compute the number of missing values for each column.
2. Compute the percentage of missing values for each column, using the length of `X_train` as the denominator.
3. Create a DataFrame named `missing_info` with two columns:
   - `Missing Count`
   - `Missing Percentage`
4. Keep only the rows where `Missing Count` is greater than 0.
5. Sort `missing_info` in descending order by `Missing Percentage`.
6. Print a message indicating that missing-value statistics are being displayed, and then display the resulting table.


In [3]:
import pandas as pd

# 1. Compute the number of missing values for each column
missing_counts = X_train.isnull().sum()

# 2. Compute the percentage of missing values
missing_percentages = (missing_counts / len(X_train)) * 100

# 3. Create the DataFrame missing_info
missing_info = pd.DataFrame({
    'Missing Count': missing_counts,
    'Missing Percentage': missing_percentages
})

# 4. Keep only rows where Missing Count > 0
missing_info = missing_info[missing_info['Missing Count'] > 0]

# 5. Sort in descending order by Missing Percentage
missing_info = missing_info.sort_values(by='Missing Percentage', ascending=False)

# 6. Print message and display table
print("Displaying missing-value statistics for X_train:")
display(missing_info)


Displaying missing-value statistics for X_train:


,Missing Count,Missing Percentage
PoolQC,1162,99.486301
MiscFeature,1122,96.061644
Alley,1094,93.664384
Fence,935,80.051370
MasVnrType,683,58.476027
FireplaceQu,547,46.832192
LotFrontage,217,18.578767
GarageType,64,5.479452
GarageYrBlt,64,5.479452
GarageFinish,64,5.479452


## Task 4.3. drop variables that have 80% or above missing variables
-CAVEAT: We drop those variables with lots of missings for this in-class demostration. In practice, you shall think clearly on if you wanted to drop those variables.

- Intention: if a column has too little information and too many missing values, it is less useful, and we could drop it
- Prompt: Work with the DataFrame `missing_info` and perform the following steps:

1. Identify all columns where `Missing Percentage` is greater than or equal to 80%.  
2. Store these column names in a list called `columns_to_drop`.
3. Drop those columns from `X_train`.
4. Print the list of dropped columns.
5. Print the new shape of `X_train`.

-**CRITICAL NOTE**: any variable that is in columns_to_drop, will be dropped in both the train set and the test set

In [4]:
# 1. Identify columns where Missing Percentage is >= 80%
# 2. Store these column names in a list called columns_to_drop
columns_to_drop = missing_info[missing_info['Missing Percentage'] >= 80].index.tolist()

# 3. Drop those columns from X_train
X_train = X_train.drop(columns=columns_to_drop)

# 4. Print the list of dropped columns
print(f"Dropped columns with >= 80% missing values: {columns_to_drop}")

# 5. Print the new shape of X_train
print(f"New shape of X_train: {X_train.shape}")


Dropped columns with >= 80% missing values: ['PoolQC', 'MiscFeature', 'Alley', 'Fence']
New shape of X_train: (1168, 76)


## Task 4.4. Split X_train into two parts: one contains all numerical variable, the other contains all non-numerical variable
- Intention: numerical and non-numerical variables shall be treated differently in handling missing
- Prompt: split X_train into X_train_num that contains all numerical variables, and X_train_not_num that contains all non-numerical variables

In [5]:
import numpy as np

# Split into numerical and non-numerical features
X_train_num = X_train.select_dtypes(include=np.number)
X_train_not_num = X_train.select_dtypes(exclude=np.number)

# Print confirmation and shapes
print("X_train split into numerical and non-numerical dataframes successfully!")
print(f"Shape of X_train_num: {X_train_num.shape}")
print(f"Shape of X_train_not_num: {X_train_not_num.shape}")


X_train split into numerical and non-numerical dataframes successfully!
Shape of X_train_num: (1168, 37)
Shape of X_train_not_num: (1168, 39)


## Task 4.5. Let's first work with the dataframe contains all numerical variable, use imputation to fill in missing values

### Task 4.5.1. Examine missing values in X_train_num
- Intention: First, let's examine what variables in X_train_num has missing values, and how many values are missing
- Prompt: Work with the DataFrame `X_train_num` and perform the following steps:

1. Compute the number of missing values for each numerical column.
2. Compute the percentage of missing values for each column, using the length of `X_train_num` as the denominator.
3. Create a DataFrame named `missing_info_num` with two columns:
   - `Missing Count`
   - `Missing Percentage`
4. Keep only the rows where `Missing Count` is greater than 0.
5. Sort `missing_info_num` in descending order by `Missing Percentage`.
6. Print a message indicating that missing-value statistics for numerical variables are being displayed, and then display the resulting table.

In [6]:
import pandas as pd

# 1. Compute the number of missing values for each numerical column
missing_counts_num = X_train_num.isnull().sum()

# 2. Compute the percentage of missing values
missing_percentages_num = (missing_counts_num / len(X_train_num)) * 100

# 3. Create the DataFrame missing_info_num
missing_info_num = pd.DataFrame({
    'Missing Count': missing_counts_num,
    'Missing Percentage': missing_percentages_num
})

# 4. Keep only rows where Missing Count > 0
missing_info_num = missing_info_num[missing_info_num['Missing Count'] > 0]

# 5. Sort in descending order by Missing Percentage
missing_info_num = missing_info_num.sort_values(by='Missing Percentage', ascending=False)

# 6. Print message and display table
print("Percentage of missing values in numerical variables of X_train_num:")
display(missing_info_num)


Percentage of missing values in numerical variables of X_train_num:


,Missing Count,Missing Percentage
LotFrontage,217,18.578767
GarageYrBlt,64,5.479452
MasVnrArea,6,0.513699


### Task 4.5.2. Use KNN to imput missing value for "GarageYrBlt" and "LotFrontage"
- Intention: GaragreYrBlt are very much likely to correlated with other features.
- Prompt: Work with the DataFrame `X_train_num` and perform the following steps:

1. Import `KNNImputer` from `sklearn.impute`.
2. Initialize a `KNNImputer` with `n_neighbors=5`.
3. Create a temporary copy of `X_train_num` named `X_train_num_temp` for imputation.
4. Fit the imputer on `X_train_num_temp` and transform it to produce an imputed array.
5. Convert the imputed array back into a DataFrame named `X_train_num_temp_df`, preserving the original column names and index.
6. Update only the columns `GarageYrBlt` and `LotFrontage` in `X_train_num` using the corresponding columns from `X_train_num_temp_df`.

In [7]:
from sklearn.impute import KNNImputer
import pandas as pd

# 1. & 2. Initialize a KNNImputer with n_neighbors=5
knn_imputer = KNNImputer(n_neighbors=5)

# 3. Create a temporary copy of X_train_num for imputation
X_train_num_temp = X_train_num.copy()

# 4. Fit the imputer and transform the data
imputed_array = knn_imputer.fit_transform(X_train_num_temp)

# 5. Convert the imputed array back into a DataFrame
X_train_num_temp_df = pd.DataFrame(
    imputed_array,
    columns=X_train_num_temp.columns,
    index=X_train_num_temp.index
)

# 6. Update only 'GarageYrBlt' and 'LotFrontage' in X_train_num
X_train_num['GarageYrBlt'] = X_train_num_temp_df['GarageYrBlt']
X_train_num['LotFrontage'] = X_train_num_temp_df['LotFrontage']

print("Successfully imputed 'GarageYrBlt' and 'LotFrontage' using KNNImputer.")


Successfully imputed 'GarageYrBlt' and 'LotFrontage' using KNNImputer.


-**CRITICAL NOTES:** The fitted knn_imputer will be used on the test set later. NEVER re-fit imputer on test set

### Task 4.5.3. Use median to fill missing values in "MasVnrArea".
- Only less than 1 percent value in MasVnrArea" were missing, so using median to impute it won't introduce too much bias. For class demonstration purpose, let's try median imputer
- Prompt: Work with the DataFrame `X_train_num` and perform the following steps:

1. Import `SimpleImputer` from `sklearn.impute`.
2. Initialize a `SimpleImputer` using the `"median"` strategy.
3. Fit the imputer on the `MasVnrArea` column of `X_train_num` and transform that column.
4. Update `X_train_num['MasVnrArea']` with the imputed values.

In [8]:
from sklearn.impute import SimpleImputer

# Initialize a SimpleImputer using the "median" strategy
median_imputer = SimpleImputer(strategy='median')

# Fit the imputer on the MasVnrArea column of X_train_num and transform that column
X_train_num['MasVnrArea'] = median_imputer.fit_transform(X_train_num[['MasVnrArea']])

print("Successfully imputed 'MasVnrArea' with its median value.")


Successfully imputed 'MasVnrArea' with its median value.


- **CRITICAL NOTES:** the fitted median_imputer will be used again in the test set. Never re-fit imputers on test set.

### Task 4.5.4. Check if we still have missing values in X_train_num
- Intention: Sanity Check
- Prompt: Work with the DataFrame `X_train_num` and perform the following steps:

1. Compute the number of remaining missing values in each column and store the result in `missing_values_num_final`.
2. Keep only the columns where the missing count is greater than 0.
3. If `missing_values_num_final` is empty, print a message indicating that no missing values remain.
4. Otherwise, print a message indicating that missing values are still present and display the list of columns with remaining missing values.

In [9]:
# 1. Compute the number of remaining missing values in each column
missing_values_num_final = X_train_num.isnull().sum()

# 2. Keep only the columns where the missing count is greater than 0
missing_values_num_final = missing_values_num_final[missing_values_num_final > 0]

# 3 & 4. Print appropriate message based on if missing values remain
if missing_values_num_final.empty:
    print("Hooray! No missing values remaining in X_train_num.")
else:
    print("Missing values are still present in the following columns:")
    display(missing_values_num_final)


Hooray! No missing values remaining in X_train_num.


## Task 4.6. Next let's work with non-numerical variables

### Task 4.6.1. Examine missing values in non-numerical variables

- **Intention:** Identify non-numerical variables that contain missing values and understand the severity of missingness before deciding how to handle them.

- **Prompt:**  
Work with the DataFrame `X_train_not_num` and perform the following steps:

1. Compute the number of missing values for each non-numerical column.
2. Compute the percentage of missing values for each column, using the length of `X_train_not_num` as the denominator.
3. Create a DataFrame named `missing_info_not_num` with two columns:
   - `Missing Count`
   - `Missing Percentage`
4. Keep only the rows where `Missing Count` is greater than 0.
5. Sort `missing_info_not_num` in descending order by `Missing Percentage`.
6. Print a message indicating that missing-value statistics for non-numerical variables are being displayed, and then display the resulting table.


In [10]:
import pandas as pd

# 1. Compute the number of missing values for each non-numerical column
missing_counts_not_num = X_train_not_num.isnull().sum()

# 2. Compute the percentage of missing values
missing_percentages_not_num = (missing_counts_not_num / len(X_train_not_num)) * 100

# 3. Create the DataFrame missing_info_not_num
missing_info_not_num = pd.DataFrame({
    'Missing Count': missing_counts_not_num,
    'Missing Percentage': missing_percentages_not_num
})

# 4. Keep only the rows where Missing Count is greater than 0
missing_info_not_num = missing_info_not_num[missing_info_not_num['Missing Count'] > 0]

# 5. Sort missing_info_not_num in descending order by Missing Percentage
missing_info_not_num = missing_info_not_num.sort_values(by='Missing Percentage', ascending=False)

# 6. Print message and display table
print("Displaying missing-value statistics for non-numerical variables in X_train_not_num:")
display(missing_info_not_num)


Displaying missing-value statistics for non-numerical variables in X_train_not_num:


,Missing Count,Missing Percentage
MasVnrType,683,58.476027
FireplaceQu,547,46.832192
GarageType,64,5.479452
GarageFinish,64,5.479452
GarageQual,64,5.479452
GarageCond,64,5.479452
BsmtQual,28,2.397260
BsmtCond,28,2.397260
BsmtExposure,28,2.397260
BsmtFinType1,28,2.397260


### Task 4.6.2. Check what are the unique values for each of the variables listed above?

- **Intention:** Review the distinct category values for each non-numerical variable that contains missing data, to understand the full set of possible categories before deciding how to impute or encode them.

- **Prompt:**  
Work with the DataFrame `X_train_not_num` and the index of `missing_info_not_num`, and perform the following steps:

1. Print a message indicating that unique values are being inspected for variables that previously had missing data.
2. Loop through each column name in `missing_info_not_num.index`.
3. For each column:
   - If the column exists in `X_train_not_num`, print the column name and display its unique values.
   - If the column does not exist in `X_train_not_num`, print a message indicating that it may have been dropped earlier.


In [11]:
print("Inspecting unique values for variables that previously had missing data:")

for col in missing_info_not_num.index:
    if col in X_train_not_num.columns:
        print(f"\nColumn: {col}")
        print(X_train_not_num[col].unique())
    else:
        print(f"\nColumn: {col} (Not found in X_train_not_num, may have been dropped earlier)")


Inspecting unique values for variables that previously had missing data:

Column: MasVnrType
[nan 'BrkFace' 'Stone' 'BrkCmn']

Column: FireplaceQu
[nan 'TA' 'Gd' 'Fa' 'Ex' 'Po']

Column: GarageType
['Attchd' nan 'Detchd' 'BuiltIn' 'Basment' '2Types' 'CarPort']

Column: GarageFinish
['RFn' nan 'Unf' 'Fin']

Column: GarageQual
['TA' nan 'Po' 'Fa' 'Gd' 'Ex']

Column: GarageCond
['TA' nan 'Po' 'Fa' 'Gd' 'Ex']

Column: BsmtQual
['TA' 'Gd' 'Fa' nan 'Ex']

Column: BsmtCond
['TA' 'Fa' nan 'Gd' 'Po']

Column: BsmtExposure
['No' 'Av' 'Gd' 'Mn' nan]

Column: BsmtFinType1
['Rec' 'Unf' 'ALQ' 'LwQ' 'BLQ' 'GLQ' nan]

Column: BsmtFinType2
['Unf' 'LwQ' 'Rec' 'BLQ' nan 'GLQ' 'ALQ']

Column: Electrical
['SBrkr' 'FuseA' 'FuseF' 'FuseP' nan]


### Task 4.6.3. Handle missing values in non-numerical variables
- **Intention:** Replace missing values in non-numerical variables with the string `"None"` so that these variables can be consistently processed during encoding.

- **Prompt:**  
Work with the DataFrame `X_train_not_num` and perform the following steps:

1. Fill all missing values in `X_train_not_num` with the string `"None"`.
2. Print a confirmation message.
3. Optionally display the updated missing-value counts to verify that no missing values remain.


- **CRITICAL NOTES** Apply the EXACT SAME procedure on the non-numerical variables in the test set later.

In [12]:
# 1. Fill all missing values in X_train_not_num with the string "None"
X_train_not_num = X_train_not_num.fillna("None")

# 2. Print a confirmation message
print("Successfully replaced all missing values in X_train_not_num with 'None'.")

# 3. Optionally display the updated missing-value counts to verify that no missing values remain
missing_counts_after = X_train_not_num.isnull().sum()
missing_remaining = missing_counts_after[missing_counts_after > 0]

if missing_remaining.empty:
    print("Verification: No missing values remain in X_train_not_num.")
else:
    print("Verification failed. Missing values still present in:")
    display(missing_remaining)


Successfully replaced all missing values in X_train_not_num with 'None'.
Verification: No missing values remain in X_train_not_num.


### Task 4.6.4. Verify if all missing values in `X_train_not_num` are handled
- **Intention:** Confirm that all missing values in the non-numerical portion of the training data have been successfully filled.

- **Prompt:**  
Work with the DataFrame `X_train_not_num` and perform the following steps:

1. Compute the number of remaining missing values in each column and store the result.
2. Check whether any columns still contain missing values.
3. If no missing values remain, print a confirmation message.
4. Otherwise, print a message indicating that missing values are still present and display the columns that contain them.


In [13]:
# 1. Compute the number of remaining missing values in each column
missing_values_not_num_final = X_train_not_num.isnull().sum()

# 2. Check whether any columns still contain missing values
missing_values_not_num_final = missing_values_not_num_final[missing_values_not_num_final > 0]

# 3. & 4. Print confirmation or error message
if missing_values_not_num_final.empty:
    print("Confirmation: No missing values remain in X_train_not_num.")
else:
    print("Missing values are still present in the following columns:")
    display(missing_values_not_num_final)


Confirmation: No missing values remain in X_train_not_num.


## Step 5. Feature engineering


## Task 5.1. In this task, we will keep working with the numerical dataset, adjusting for skewness and outliner

### Task 5.1.1. Check for variable skewness
- Intention: We need to have a basic understanding on data skewness
- Prompt: working with the dataset "X_train_num", calculate the skewness for each variable, sorted descending, and display them

In [14]:
# Calculate skewness for each numerical variable
skewness = X_train_num.skew().sort_values(ascending=False)

# Display the sorted skewness values
print("Skewness of numerical variables (descending):")
display(skewness)


Skewness of numerical variables (descending):


MiscVal          22.053579
PoolArea         14.396098
LotArea          11.958088
3SsnPorch         9.833911
LowQualFinSF      9.199456
KitchenAbvGr      4.445055
BsmtFinSF2        4.217895
ScreenPorch       4.090393
BsmtHalfBath      4.005786
EnclosedPorch     3.163946
OpenPorchSF       2.331890
MasVnrArea        2.294117
LotFrontage       2.151989
BsmtFinSF1        1.862132
TotalBsmtSF       1.723881
WoodDeckSF        1.587330
MSSubClass        1.438804
GrLivArea         1.425139
1stFlrSF          1.422162
BsmtUnfSF         0.910628
2ndFlrSF          0.801209
OverallCond       0.733180
TotRmsAbvGrd      0.687336
HalfBath          0.651296
Fireplaces        0.637526
BsmtFullBath      0.612032
BedroomAbvGr      0.289882
MoSold            0.241761
OverallQual       0.196575
GarageArea        0.109110
YrSold            0.082593
FullBath          0.038169
Id                0.000920
GarageCars       -0.389732
YearRemodAdd     -0.508067
GarageYrBlt      -0.605232
YearBuilt        -0.623404
d

### Task 5.1.2. Adjust skewness with log transformation
- Intention: we need to adjust for skewness of the data so that our model will learn the patterns, rather than chasing outliers
- Prompt: Create a copy of the dataframe `X_train_num_adjusted`, then apply log transformation (log1p) on variables that skewness > 1

In [15]:
import numpy as np

# Create a copy of the dataframe
X_train_num_adjusted = X_train_num.copy()

# Identify variables with skewness > 1
highly_skewed_vars = skewness[skewness > 1].index

# Apply log1p transformation to these variables
for col in highly_skewed_vars:
    X_train_num_adjusted[col] = np.log1p(X_train_num_adjusted[col].astype('float'))

print(f"Applied log1p transformation to {len(highly_skewed_vars)} highly skewed variables.")


Applied log1p transformation to 19 highly skewed variables.


### Task 5.1.3. Check skewness again after the transformation

In [16]:
# Calculate skewness for each numerical variable after transformation
skewness_adjusted = X_train_num_adjusted.skew().sort_values(ascending=False)

# Display the sorted skewness values
print("Skewness of numerical variables after log transformation (descending):")
display(skewness_adjusted)


Skewness of numerical variables after log transformation (descending):


PoolArea         13.874124
3SsnPorch         7.510573
LowQualFinSF      7.447547
MiscVal           5.004489
BsmtHalfBath      3.893541
KitchenAbvGr      3.722592
ScreenPorch       3.062578
BsmtFinSF2        2.561997
EnclosedPorch     2.179302
BsmtUnfSF         0.910628
2ndFlrSF          0.801209
OverallCond       0.733180
TotRmsAbvGrd      0.687336
HalfBath          0.651296
Fireplaces        0.637526
BsmtFullBath      0.612032
MasVnrArea        0.452588
BedroomAbvGr      0.289882
MSSubClass        0.255827
MoSold            0.241761
OverallQual       0.196575
WoodDeckSF        0.181798
GarageArea        0.109110
YrSold            0.082593
FullBath          0.038169
1stFlrSF          0.026976
GrLivArea         0.007943
Id                0.000920
LotArea          -0.012599
OpenPorchSF      -0.064149
GarageCars       -0.389732
YearRemodAdd     -0.508067
GarageYrBlt      -0.605232
BsmtFinSF1       -0.622601
YearBuilt        -0.623404
LotFrontage      -0.700318
TotalBsmtSF      -5.274632
d

## Task 5.2. Next, let's make non-numerical variables more meaningful

### Task 5.2.1. After some investigation, we found that all the non-numerical variables can be treated as categorical variables (i.e., no pure text descriptions). Let's first convert all of them into categorical variables.
- Intention: categorical datatype is easier to dealt with
- Prompt: convert all "object" dtype to "category" dtype in X_train_non_num

In [17]:
print("Converting 'object' dtype columns to 'category' dtype in X_train_not_num...")

for col in X_train_not_num.columns:
    if X_train_not_num[col].dtype == 'object':
        X_train_not_num[col] = X_train_not_num[col].astype('category')
        print(f"  - Converted '{col}' to 'category'")

# print("Conversion complete. Displaying dtypes of X_train_not_num:")
display(X_train_not_num.dtypes)


Converting 'object' dtype columns to 'category' dtype in X_train_not_num...
  - Converted 'MSZoning' to 'category'
  - Converted 'Street' to 'category'
  - Converted 'LotShape' to 'category'
  - Converted 'LandContour' to 'category'
  - Converted 'Utilities' to 'category'
  - Converted 'LotConfig' to 'category'
  - Converted 'LandSlope' to 'category'
  - Converted 'Neighborhood' to 'category'
  - Converted 'Condition1' to 'category'
  - Converted 'Condition2' to 'category'
  - Converted 'BldgType' to 'category'
  - Converted 'HouseStyle' to 'category'
  - Converted 'RoofStyle' to 'category'
  - Converted 'RoofMatl' to 'category'
  - Converted 'Exterior1st' to 'category'
  - Converted 'Exterior2nd' to 'category'
  - Converted 'MasVnrType' to 'category'
  - Converted 'ExterQual' to 'category'
  - Converted 'ExterCond' to 'category'
  - Converted 'Foundation' to 'category'
  - Converted 'BsmtQual' to 'category'
  - Converted 'BsmtCond' to 'category'
  - Converted 'BsmtExposure' to 'catego

MSZoning         category
Street           category
LotShape         category
LandContour      category
Utilities        category
LotConfig        category
LandSlope        category
Neighborhood     category
Condition1       category
Condition2       category
BldgType         category
HouseStyle       category
RoofStyle        category
RoofMatl         category
Exterior1st      category
Exterior2nd      category
MasVnrType       category
ExterQual        category
ExterCond        category
Foundation       category
BsmtQual         category
BsmtCond         category
BsmtExposure     category
BsmtFinType1     category
BsmtFinType2     category
Heating          category
HeatingQC        category
CentralAir       category
Electrical       category
KitchenQual      category
Functional       category
FireplaceQu      category
GarageType       category
GarageFinish     category
GarageQual       category
GarageCond       category
PavedDrive       category
SaleType         category
SaleConditio

### Task 5.2.2. We need to better understand what are the values for each of the categorical variable
- Intention: by looking at the values each variable take, we can better decide what variables can benefit from ordinal encoding, and what variables benefit from one-hot encoding
- Prompt: Check what are the unique values for each of the variables in X_train_non_num

In [18]:
for col in X_train_not_num.columns:
    print(f"Unique values in {col}:")
    print(X_train_not_num[col].unique())
    print("-" * 40)


Unique values in MSZoning:
['RL', 'RM', 'FV', 'RH', 'C (all)']
Categories (5, object): ['C (all)', 'FV', 'RH', 'RL', 'RM']
----------------------------------------
Unique values in Street:
['Pave', 'Grvl']
Categories (2, object): ['Grvl', 'Pave']
----------------------------------------
Unique values in LotShape:
['Reg', 'IR1', 'IR2', 'IR3']
Categories (4, object): ['IR1', 'IR2', 'IR3', 'Reg']
----------------------------------------
Unique values in LandContour:
['Lvl', 'Bnk', 'HLS', 'Low']
Categories (4, object): ['Bnk', 'HLS', 'Low', 'Lvl']
----------------------------------------
Unique values in Utilities:
['AllPub', 'NoSeWa']
Categories (2, object): ['AllPub', 'NoSeWa']
----------------------------------------
Unique values in LotConfig:
['Inside', 'Corner', 'CulDSac', 'FR2', 'FR3']
Categories (5, object): ['Corner', 'CulDSac', 'FR2', 'FR3', 'Inside']
----------------------------------------
Unique values in LandSlope:
['Gtl', 'Mod', 'Sev']
Categories (3, object): ['Gtl', 'Mod', 

### Task 5.2.3. After some investigation, we decided to split the `X_train_not_num.columns` into two parts, one for variables with ordinal relationship, one for variables without ordinal relationship.

In [19]:
# Manually specify categorical variables to treat as ordinal
manual_ordinal_cols = [
    'LotShape',
    'LandSlope',
    'ExterQual',
    'ExterCond',
    'BsmtQual',
    'BsmtCond',
    'BsmtExposure',
    'BsmtFinType1',
    'BsmtFinType2',
    'HeatingQC',
    'KitchenQual',
    'Functional',
    'FireplaceQu',
    'GarageFinish',
    'GarageQual',
    'GarageCond',
    'PavedDrive'
]

# Keep only those that are actually present in X_train_not_num
ordinal_col_names = [col for col in manual_ordinal_cols if col in X_train_not_num.columns]

# Split into ordinal vs nominal categorical dataframes
X_train_non_num_ordinal = X_train_not_num[ordinal_col_names].copy()
X_train_non_num_nominal = X_train_not_num.drop(columns=ordinal_col_names)


print("Finished splitting non-numerical features into ordinal and nominal:")
print(f"  X_train_non_num_ordinal shape: {X_train_non_num_ordinal.shape}")
print(f"  X_train_non_num_nominal shape: {X_train_non_num_nominal.shape}")
print(f"  Ordinal columns used: {ordinal_col_names}")


Finished splitting non-numerical features into ordinal and nominal:
  X_train_non_num_ordinal shape: (1168, 17)
  X_train_non_num_nominal shape: (1168, 22)
  Ordinal columns used: ['LotShape', 'LandSlope', 'ExterQual', 'ExterCond', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'HeatingQC', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive']


### Task 5.2.4. Identify ordinal categorical variables, and manually encoding them as integer


In [20]:
# Manual ordinal encoding for Ames Housing categoricals

# Create a copy to store the encoded ordinal features
X_train_ordinal_encoded = X_train_non_num_ordinal.copy()

# Quality/Condition scales (Po < Fa < TA < Gd < Ex)
quality_scale = {
    "None": 0, "Po": 1, "Fa": 2, "TA": 3, "Gd": 4, "Ex": 5
}

X_train_ordinal_encoded['ExterQual']  = X_train_ordinal_encoded['ExterQual'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['ExterCond']  = X_train_ordinal_encoded['ExterCond'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['BsmtQual']   = X_train_ordinal_encoded['BsmtQual'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['BsmtCond']   = X_train_ordinal_encoded['BsmtCond'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['HeatingQC']  = X_train_ordinal_encoded['HeatingQC'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['KitchenQual']= X_train_ordinal_encoded['KitchenQual'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['GarageQual'] = X_train_ordinal_encoded['GarageQual'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['GarageCond'] = X_train_ordinal_encoded['GarageCond'].map(quality_scale).astype('int64')
X_train_ordinal_encoded['FireplaceQu']= X_train_ordinal_encoded['FireplaceQu'].map(quality_scale).astype('int64')

# Basement finish types
bsmt_fin_scale = {
    "None": 0, "Unf": 1, "LwQ": 2, "Rec": 3, "BLQ": 4, "ALQ": 5, "GLQ": 6
}

X_train_ordinal_encoded['BsmtFinType1'] = X_train_ordinal_encoded['BsmtFinType1'].map(bsmt_fin_scale).astype('int64')
X_train_ordinal_encoded['BsmtFinType2'] = X_train_ordinal_encoded['BsmtFinType2'].map(bsmt_fin_scale).astype('int64')

# Basement exposure (None < No < Mn < Av < Gd)
bsmt_exposure_scale = {
    "None": 0, "No": 1, "Mn": 2, "Av": 3, "Gd": 4
}

X_train_ordinal_encoded['BsmtExposure'] = X_train_ordinal_encoded['BsmtExposure'].map(bsmt_exposure_scale).astype('int64')

# Garage finish (None < Unf < RFn < Fin)
garage_finish_scale = {
    "None": 0, "Unf": 1, "RFn": 2, "Fin": 3
}

X_train_ordinal_encoded['GarageFinish'] = X_train_ordinal_encoded['GarageFinish'].map(garage_finish_scale).astype('int64')

# Lot shape (most irregular → most regular)
lot_shape_scale = {
    "None": 0, "IR3": 1, "IR2": 2, "IR1": 3, "Reg": 4
}

X_train_ordinal_encoded['LotShape'] = X_train_ordinal_encoded['LotShape'].map(lot_shape_scale).astype('int64')

# Land slope (Sev < Mod < Gtl)
land_slope_scale = {
    "None": 0, "Sev": 1, "Mod": 2, "Gtl": 3
}

X_train_ordinal_encoded['LandSlope'] = X_train_ordinal_encoded['LandSlope'].map(land_slope_scale).astype('int64')

# Paved driveway (N < P < Y)
paved_drive_scale = {
    "None": 0, "N": 1, "P": 2, "Y": 3
}

X_train_ordinal_encoded['PavedDrive'] = X_train_ordinal_encoded['PavedDrive'].map(paved_drive_scale).astype('int64')

# Functional (Sev → major → minor → typical)
functional_scale = {
    "None": 0, "Sev": 1, "Maj2": 2, "Maj1": 3,
    "Mod": 4, "Min2": 5, "Min1": 6, "Typ": 7
}

X_train_ordinal_encoded['Functional'] = X_train_ordinal_encoded['Functional'].map(functional_scale).astype('int64')

print("Ordinal encoding applied to X_train_non_num_ordinal and stored in X_train_ordinal_encoded.")
print(f"Shape of X_train_ordinal_encoded: {X_train_ordinal_encoded.shape}")
display(X_train_ordinal_encoded.head())

Ordinal encoding applied to X_train_non_num_ordinal and stored in X_train_ordinal_encoded.
Shape of X_train_ordinal_encoded: (1168, 17)


,LotShape,LandSlope,ExterQual,ExterCond,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinType2,HeatingQC,KitchenQual,Functional,FireplaceQu,GarageFinish,GarageQual,GarageCond,PavedDrive
254,4,3,3,4,3,3,1,3,1,3,3,7,0,2,3,3,3
1066,3,3,4,3,4,3,1,1,1,4,3,7,3,2,3,3,3
638,4,3,3,3,2,3,1,1,1,4,3,7,0,0,0,0,2
799,4,3,3,3,4,3,1,5,1,5,4,7,3,1,3,3,3
380,4,3,3,3,3,3,1,2,1,3,4,7,4,1,3,3,3


### Task 5.2.5. put all variables in `X_train_non_num` but do not have ordinal relationship into a new dataframe `X_train_non_num_nominal`
- Intention: Create a dedicated DataFrame to handle categorical variables that does not have clear ordinal relationship
- Prompt: Work with the DataFrame `X_train_not_num` and the list `ordinal_col_names`, and create a new DataFrame `X_train_non_num_nominal` by selecting all columns in `X_train_not_num` that are not listed in `ordinal_col_names`.

In [21]:
nominal_cols = [col for col in X_train_not_num.columns if col not in ordinal_col_names]
X_train_non_num_nominal = X_train_not_num[nominal_cols].copy()

print("Created X_train_non_num_nominal DataFrame successfully!")
print(f"Shape of X_train_non_num_nominal: {X_train_non_num_nominal.shape}")


Created X_train_non_num_nominal DataFrame successfully!
Shape of X_train_non_num_nominal: (1168, 22)


### Task 5.2.6. Use a one-hot encoder to convert all nominal categorical variables (i.e., variables without any meaningful ordinal relationship) into dummy variables, ensuring the encoding is consistent and safe for downstream model training.
- Intention:Nominal variables do not have a natural order, so integer encoding would incorrectly impose structure. One-hot encoding expands each nominal feature into a series of binary indicator columns, preserving the categorical information without implying hierarchy.
We apply one-hot encoding to ensure models—especially linear models and tree-based models—treat these variables correctly.
- Prompt:working with the dataset `X_train_non_num_nominal`,use sklearn.preprocessing.OneHotEncoder to one-hot encode all columns in `X_train_nominal`.

Create a OneHotEncoder with the following requirements:

1. sparse_output=False (so the result is a dense DataFrame)
2. handle_unknown='ignore' (so the encoder can be applied safely to the test set even if unseen categories appear)
3. Fit the encoder on `X_train_nominal`, then transform it to produce a DataFrame called `X_train_nominal_encoded`.
4. Ensure the output is a pandas DataFrame with correct column names, derived from encoder.get_feature_names_out().
5. Print:The shape and head of `X_train_nominal_encoded`
 and The list of generated feature names

In [22]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

# Initialize the OneHotEncoder
onehot_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

# Fit and transform the nominal data
encoded_array = onehot_encoder.fit_transform(X_train_non_num_nominal)

# Get feature names
feature_names = onehot_encoder.get_feature_names_out(X_train_non_num_nominal.columns)

# Create the encoded DataFrame, preserving the original index
X_train_nominal_encoded = pd.DataFrame(
    encoded_array,
    columns=feature_names,
    index=X_train_non_num_nominal.index
)

# Print requested information
print(f"Shape of X_train_nominal_encoded: {X_train_nominal_encoded.shape}")
print("List of generated feature names:")
print(feature_names.tolist())
display(X_train_nominal_encoded.head())


Shape of X_train_nominal_encoded: (1168, 162)
List of generated feature names:
['MSZoning_C (all)', 'MSZoning_FV', 'MSZoning_RH', 'MSZoning_RL', 'MSZoning_RM', 'Street_Grvl', 'Street_Pave', 'LandContour_Bnk', 'LandContour_HLS', 'LandContour_Low', 'LandContour_Lvl', 'Utilities_AllPub', 'Utilities_NoSeWa', 'LotConfig_Corner', 'LotConfig_CulDSac', 'LotConfig_FR2', 'LotConfig_FR3', 'LotConfig_Inside', 'Neighborhood_Blmngtn', 'Neighborhood_Blueste', 'Neighborhood_BrDale', 'Neighborhood_BrkSide', 'Neighborhood_ClearCr', 'Neighborhood_CollgCr', 'Neighborhood_Crawfor', 'Neighborhood_Edwards', 'Neighborhood_Gilbert', 'Neighborhood_IDOTRR', 'Neighborhood_MeadowV', 'Neighborhood_Mitchel', 'Neighborhood_NAmes', 'Neighborhood_NPkVill', 'Neighborhood_NWAmes', 'Neighborhood_NoRidge', 'Neighborhood_NridgHt', 'Neighborhood_OldTown', 'Neighborhood_SWISU', 'Neighborhood_Sawyer', 'Neighborhood_SawyerW', 'Neighborhood_Somerst', 'Neighborhood_StoneBr', 'Neighborhood_Timber', 'Neighborhood_Veenker', 'Conditi

,MSZoning_C (all),MSZoning_FV,MSZoning_RH,MSZoning_RL,MSZoning_RM,Street_Grvl,Street_Pave,LandContour_Bnk,LandContour_HLS,LandContour_Low,...,SaleType_ConLw,SaleType_New,SaleType_Oth,SaleType_WD,SaleCondition_Abnorml,SaleCondition_AdjLand,SaleCondition_Alloca,SaleCondition_Family,SaleCondition_Normal,SaleCondition_Partial
254,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
1066,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
638,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
799,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
380,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0


### Task 5.2.7.Combine the encoded ordinal dataset and the encoded nominal dataset back together
- Intention: now that the non-numerical variables have been encoded as numbers, we should combine them back together
- Prompt: concat the df `X_train_ordinal_encoded` and the df `X_train_nominal_encoded` back together, name the new dataframe as `X_train_non_num_encoded`. Print the shape of the new dataframe

In [23]:
X_train_non_num_encoded = pd.concat([X_train_ordinal_encoded, X_train_nominal_encoded], axis=1)

print("DataFrames X_train_ordinal_encoded and X_train_nominal_encoded merged successfully!")
print(f"Shape of X_train_non_num_encoded: {X_train_non_num_encoded.shape}")
display(X_train_non_num_encoded.head())

DataFrames X_train_ordinal_encoded and X_train_nominal_encoded merged successfully!
Shape of X_train_non_num_encoded: (1168, 179)


,LotShape,LandSlope,ExterQual,ExterCond,BsmtQual,BsmtCond,BsmtExposure,BsmtFinType1,BsmtFinType2,HeatingQC,...,SaleType_ConLw,SaleType_New,SaleType_Oth,SaleType_WD,SaleCondition_Abnorml,SaleCondition_AdjLand,SaleCondition_Alloca,SaleCondition_Family,SaleCondition_Normal,SaleCondition_Partial
254,4,3,3,4,3,3,1,3,1,3,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
1066,3,3,4,3,4,3,1,1,1,4,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
638,4,3,3,3,2,3,1,1,1,4,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
799,4,3,3,3,4,3,1,5,1,5,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
380,4,3,3,3,3,3,1,2,1,3,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0


### Task 5.2.8. Double check that all cells in the dataframe `X_train_non_num_encoded` are numerical value now

In [24]:
import pandas as pd

# Check the distribution of data types
print("Data types in X_train_non_num_encoded:")
print(X_train_non_num_encoded.dtypes.value_counts())

# Programmatically verify if all are numeric
all_numeric = all(pd.api.types.is_numeric_dtype(dtype) for dtype in X_train_non_num_encoded.dtypes)
print(f"\nAre all columns numerical? {all_numeric}")


Data types in X_train_non_num_encoded:
float64    162
int64       17
Name: count, dtype: int64

Are all columns numerical? True


### Task 5.3. Concat the dataframe "X_train_num_adjusted" and the dataframe "X_train_non_num_encoded" back together, name the new dataframe as `X_train_ready`

In [25]:
X_train_ready = pd.concat([X_train_num_adjusted, X_train_non_num_encoded], axis=1)

print("DataFrames X_train_num_adjusted and X_train_non_num_encoded concatenated successfully!")
print(f"Shape of X_train_ready: {X_train_ready.shape}")
display(X_train_ready.head())

DataFrames X_train_num_adjusted and X_train_non_num_encoded concatenated successfully!
Shape of X_train_ready: (1168, 216)


,Id,MSSubClass,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,BsmtFinSF1,...,SaleType_ConLw,SaleType_New,SaleType_Oth,SaleType_WD,SaleCondition_Abnorml,SaleCondition_AdjLand,SaleCondition_Alloca,SaleCondition_Family,SaleCondition_Normal,SaleCondition_Partial
254,255,3.044522,4.262680,9.036106,5,6,1957,1957,0.000000,6.827629,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
1066,1067,4.110874,4.094345,8.966739,6,7,1993,1994,0.000000,0.000000,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
638,639,3.433987,4.219508,9.080004,5,7,1910,1950,0.000000,0.000000,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
799,800,3.931826,4.110874,8.881975,5,7,1937,1950,5.533389,6.345636,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0
380,381,3.931826,3.931826,8.517393,5,6,1924,1950,0.000000,5.389072,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0


## Task 5.4. **CRITICALLY IMPORTANT**:
## **Document and Apply Best Practices for Preprocessing the Test Set**

Before you preprocess your test set, you must **review and clearly document** every preprocessing step you applied to the training set, including:

1. Which steps involved fitting (e.g., imputers, encoders, scalers)

2. Which steps did NOT involve fitting (e.g., log transforms, splits, merges)

3. The exact order in which all steps were applied

This documentation becomes the **“blueprint”** for how the test set must be processed.

**When processing the test set:**

- For fitted transformations:
  - Use the already-fitted object and call `.transform().`
  - Do not refit anything on the test set under any circumstances.

- For non-fitted transformations:
  - Apply the exact same operations, with the same parameters and in the same order.

- Maintain identical order of operations as used on the training set.

- Ensure the final training and test matrices have identical columns, in the same order.

This practice prevents data leakage, preserves the integrity of your evaluation, and matches real-world deployment where future data cannot influence preprocessing.

## Preprocessing Blueprint for Test Set

This document outlines the exact preprocessing steps applied to the training set (`X_train`) and serves as the blueprint for processing the test set (`X_test`). It categorizes steps by whether they involved fitting a transformer (which must *not* be refitted on the test set) and lists them in chronological order of application.

### Order of Operations:

1.  **Initial Data Splitting** (It is done already. Ever you need to do this again for some reason, keep the random_state the same, it will give you exactly the same result)
    *   `X = raw_data.drop('SalePrice', axis=1)`
    *   `y = raw_data['SalePrice']`
    *   `X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)`

2.  **Drop Columns with High Missing Percentage** (You shall not checking the percentage of missing again. Instead, drop exactly what you dropped in the train set)
    *   Identified columns: `['PoolQC', 'MiscFeature', 'Alley', 'Fence']`

3.  **Split `X_train` into Numerical and Non-Numerical DataFrames**
    *   `X_train_num = X_train.select_dtypes(include=np.number)`
    *   `X_train_not_num = X_train.select_dtypes(exclude=np.number)`

4.  **Handle Missing Numerical Values**
    *   **Impute `GarageYrBlt` and `LotFrontage` using KNNImputer**
        *   `knn_imputer = KNNImputer(n_neighbors=5)` was **fitted** on `X_train_num_temp` (a copy of `X_train_num`)---DO NOT FIT IT AGAIN
        *   The fitted `knn_imputer` was then used to transform and update `GarageYrBlt` and `LotFrontage` in `X_train_num`---DO NOT FIT IT AGAIN
    *   **Impute `MasVnrArea` using SimpleImputer (median strategy)**
        *   `median_imputer = SimpleImputer(strategy='median')` was **fitted** on `X_train_num[['MasVnrArea']]`---DO NOT FIT IT AGAIN
        *   The fitted `median_imputer` was then used to transform and update `MasVnrArea` in `X_train_num`---DO NOT FIT IT AGAIN

5.  **Apply Log1p Transformation to Highly Skewed Numerical Variables**
    * A *List* `highly_skewed_vars` was made identifying columns need log transformation
    *   A *copy* of `X_train_num` was created: `X_train_num_adjusted = X_train_num.copy()`.
    *   `np.log1p` was applied to these highly skewed columns in `X_train_num_adjusted`, ensuring explicit casting to `float` first.

6.  **Handle Missing Non-Numerical Values**
    *   Impute specified `Garage` and `Bsmt` related columns with `'None'` (e.g., 'GarageQual', 'BsmtQual', etc.) in `X_train_not_num`.
    *   Impute `Electrical` with its mode in `X_train_not_num`.

7.  **Convert 'object' dtype to 'category' dtype in `X_train_not_num`**
    *   Iterated through `X_train_not_num` and converted `object` columns to `category`.

8.  **Split Non-Numerical into Ordinal and Nominal DataFrames**
    *   `X_train_non_num_ordinal = X_train_not_num[ordinal_col_names].copy()`
    *   `X_train_non_num_nominal = X_train_not_num[nominal_col_names].copy()`

9.  **Encode Ordinal Categorical Variables** (Did NOT involve fitting, but used predefined mappings)
    *   Created `X_train_ordinal_encoded` by mapping categories in `X_train_non_num_ordinal` to integers (0 for 'None'/NaN, 1 onwards for ordered categories) based on `ordinal_category_details`.
    *   Each column was explicitly cast to `int64` after mapping.
    * *Task5.2.3-5.2.4* were the mannual coding part. the test set pre-processing shall follow the EXACTLY the same encoding procedure.

10. **One-Hot Encode Nominal Categorical Variables**
    *   `onehot_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')` was **fitted** on `X_train_non_num_nominal`.
    *   The fitted `onehot_encoder` was then used to transform  (DO NOT FIT IT AGAIN)`X_train_non_num_nominal` into `X_train_nominal_encoded`.

11. **Combine Encoded Non-Numerical Features**
    *   `X_train_non_num_encoded = pd.concat([X_train_ordinal_encoded, X_train_nominal_encoded], axis=1)`

12. **Combine All Processed Features**
    *   `X_train_ready = pd.concat([X_train_num_adjusted, X_train_non_num_encoded], axis=1)`

### Instructions for Test Set Processing:
* **A OVERALL PRINCIPLE**: You do not "learn" anything from the test set.
*   **Crucially**, for steps involved fitting (i.e,`fit_transform()`), you must use the *already fitted* `knn_imputer`, `median_imputer`, and `onehot_encoder` objects. Call their `.transform()` method, *NEVER* `.fit_transform()` or `.fit()` on the test set.
*   For all other steps, apply the exact same operations in the same order, with the same parameters, to the `X_test` data.
*   Ensure the final processed test set (`X_test_ready`) has the identical columns in the identical order as `X_train_ready`.
*  This is the practice that help you see how each and every step is being done to pre-process the data. In industry standard operation, you will conceptually do all of these, but wrap them into a data pipeline.

## Task 6. Preprocess the test set following the training blueprint

In this task, you will take everything you learned from preprocessing the training data and apply the **exact same sequence of operations** to the test set.

Your goal is to create a final test dataset, `X_test_ready`, that is fully aligned with `X_train_ready` and can be used directly for model evaluation.

### Deliverables

1. **A step-by-step implementation of data preprocessing on the test set**

   - Start from the existing `X_test` created in Task 4.1 (no need to resplit the data).
   - Reuse all **fitted objects and decisions** from the training stage, including (but not limited to):
     - The list of columns to drop (e.g., columns with ≥ 40% missing values in the training data).
     - Any imputers you fitted (e.g., `knn_imputer`, `median_imputer`).
     - Any mappings for ordinal variables (e.g., quality ratings).
     - Any encoders used for nominal variables (e.g., one-hot encoder).
   - Apply the **same operations in the same order** as you did for `X_train`, but using:
     - `transform` instead of `fit_transform` for sklearn transformers, and  
     - the same mapping dictionaries / column selections you constructed from the training data.
   - Clearly structure and comment your code so that each step in the test-set preprocessing corresponds to a step you previously applied to the training set.

2. **A preprocessed test set `X_test_ready` suitable for model training and validation**

   - `X_test_ready` must be preprocessed following the **exact same sequence of operations** as `X_train_ready`.
   - `X_test_ready` must have:
     - The **same number of columns** as `X_train_ready`, and  
     - The **exact same column names** (including order, if possible).
   - At the end, print out:
     - The shape of `X_test_ready`.
     - A quick check (e.g., set comparison) that the columns of `X_test_ready` and `X_train_ready` match.

The key idea: treat your training preprocessing as a **blueprint/pipeline**, and reproduce that blueprint on the test set without refitting anything on `X_test`.


### Task 6.1. drop the coloumns in the list ` columns_to_drop` on the X_test

### Task 6.2. Split X_test into Numerical and Non-Numerical DataFrames (Did NOT involve fitting)

### Task 6.3. handle missing numerical values

### Task 6.3.1. Impute GarageYrBlt and LotFrontage using KNNImputer (INVOLVED FITTING)

### Task 6.3.2. Impute 'MasVnrArea' with the fitted midian imputer

### Task 6.3.2. Check if we still have missing values in X_test_num

### Task 6.4. Handling missing categorical variables

### Task 6.4.1. Filling all missing values in X_test_not_num with 'None'

### Task 6.5. Feature engineering

### Task 6.5.1. Adjust for skewness on the variables in the list 'highly_skewed'

### Task 6.5.2. Convert all variable type 'dtype' in X_test_not_num to 'category'

### Task 6.5.3. Split the X_test_not_num.columns into two parts, one for variables with ordinal relationship, one for variables without ordinal relationship.

### Task 6.5.4. Identify ordinal categorical variables, and manually encoding them as integer

### Task 6.5.5. put all variables in X_test_non_num but do not have ordinal relationship into a new dataframe X_test_non_num_nominal

### Task 6.5.6. Use the one-hot encoder we fitted in the train set to convert all nominal categorical variables (i.e., variables without any meaningful ordinal relationship) into dummy variable

### Task 6.5.7. Combine the encoded ordinal dataset and the encoded nominal dataset back together

### Task 6.5.8. Double check if all the data in X_test_non_num_ecoded are numerical now.

### Task 6.5.9. Concat the dataframe "X_test_num_adjusted" and the dataframe "X_test_non_num_encoded" back together, name the new dataframe as X_test_ready

### Task 6.5.10. Verify that X_train_ready and X_test_ready contain the same columns, with identical names and in the same order.